cli/cli_test.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"flag"
+	"github.com/cloudwan/gohan/sync/etcdv3"
+	"github.com/codegangsta/cli"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sync"
+	"time"
+)
+
+const masterTTL = 10
+
+func getContextWithConfig(configFile string) *cli.Context {
+	f := cli.StringFlag{Name: "config-file", Value: configFile}
+	set := flag.NewFlagSet("", flag.ContinueOnError)
+	f.Apply(set)
+	return cli.NewContext(nil, set, &cli.Context{})
+}
+
+var _ = Describe("CLI", func() {
+	Describe("Post migration subcommand wrapper tests", func() {
+		It("Should lock - migrationsSubCommand wrapper", func() {
+			const (
+				configPath = "../tests/test_etcd.yaml"
+				etcdServer = "http://127.0.0.1:2379"
+			)
+
+			x := 0
+			mutex := sync.RWMutex{}
+
+			change := func(context *cli.Context) {
+				mutex.Lock()
+				defer mutex.Unlock()
+				x++
+			}
+
+			read := func() int {
+				mutex.RLock()
+				defer mutex.RUnlock()
+				return x
+			}
+
+			s, err := etcdv3.NewSync([]string{etcdServer}, time.Second)
+			Expect(err).ToNot(HaveOccurred())
+
+			block := func(c chan int) {
+				s.Lock(syncMigrationsPath, true)
+				defer s.Unlock(syncMigrationsPath)
+				c <- 0
+				time.Sleep(masterTTL / 2 * time.Second)
+			}
+
+			wrapped := migrationSubcommandWrapper(change)
+			context := getContextWithConfig(configPath)
+			Expect(context.String("config-file")).To(Equal(configPath))
+
+			c := make(chan int)
```

Comment:
what you need here is a WaitGroup (cf. https://golang.org/pkg/sync/#WaitGroup), not a channel